stl/inc/span

```diff
@@ -0,0 +1,758 @@
+    constexpr explicit _Span_extent_type(const _Span_extent_type<_OtherExtent> _Ext) noexcept {
+        static_assert(_OtherExtent == _Extent || _OtherExtent == dynamic_extent,
+            "Mismatch between static extent and size of initializing data");
+#if _CONTAINER_DEBUG_LEVEL > 0
+        if constexpr (_OtherExtent == dynamic_extent) {
+            _STL_VERIFY(_Ext._Size() == _Extent, "Mismatch between static extent and size of initializing data");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    _NODISCARD constexpr size_t size() const noexcept {
+        return _Extent;
+    }
+};
+
+template <>
+struct _Span_extent_type<dynamic_extent> {
+    constexpr _Span_extent_type() noexcept = default;
+
+    constexpr explicit _Span_extent_type(const size_t _Size) noexcept : _Mysize(_Size) {}
+
+    template <size_t _OtherExtent>
+    constexpr explicit _Span_extent_type(const _Span_extent_type<_OtherExtent> _Ext) noexcept : _Mysize(_Ext._Size()) {}
+
+    _NODISCARD constexpr size_t size() const noexcept {
+        return _Mysize;
+    }
+
+private:
+    size_t _Mysize{0};
+};
+
+// STRUCT TEMPLATE _Is_span
+template <class>
+struct _Is_span : false_type {};
+
+template <class _Ty, size_t _Extent>
+struct _Is_span<span<_Ty, _Extent>> : true_type {};
+
+// STRUCT TEMPLATE _Is_std_array
+template <class>
+struct _Is_std_array : false_type {};
+
+template <class _Ty, size_t _Extent>
+struct _Is_std_array<array<_Ty, _Extent>> : true_type {};
+
+// STRUCT TEMPLATE _Is_span_convertible_range
+template <class _Range, class _Ty>
+struct _Is_span_convertible_range
+    : bool_constant<is_convertible_v<remove_pointer_t<decltype(_STD data(_STD declval<_Range&>()))> (*)[], _Ty (*)[]>> {
+};
+
+// STRUCT TEMPLATE _Has_range_interface
+template <class, class = void>
+struct _Has_range_interface : false_type {};
+
+template <class _Range>
+struct _Has_range_interface<_Range,
+    void_t<decltype(_STD data(_STD declval<_Range&>())), decltype(_STD size(_STD declval<_Range&>()))>> : true_type {};
+
+#ifdef __cpp_lib_concepts
+// clang-format off
+template<class _It, class _Ty>
+concept _Is_span_compatible_iterator = contiguous_iterator<_It>
+                                    && is_convertible_v<remove_reference_t<iter_reference_t<_It>>(*)[], _Ty(*)[]>;
+
+template<class _It, class _Sentinel>
+concept _Is_span_compatible_sentinel = sized_sentinel_for<_Sentinel, _It>
+                                    && !is_convertible_v<_Sentinel, size_t>;
+
+template <class _Range, class _Ty>
+concept _Is_span_compatible_range = !is_array_v<remove_cvref_t<_Range>>
+                                 && !_Is_span<remove_cvref_t<_Range>>::value
+                                 && !_Is_std_array<remove_cvref_t<_Range>>::value
+                                 && ranges::contiguous_range<_Range>
+                                 && ranges::sized_range<_Range>
+                                 && (ranges::forward_range<_Range> || is_const_v<_Ty>)

```

Comment:
That shows up as a `Precondition` rather than a `Constraint` should we still leave it here, or should we `static_assert` it in the constructor?

Reply:
Ping for the language experts

Reply:
Done, I dont know how I missed the clause...